### DISCUSSION TITLE
Ship Map Editor

### DISCUSSION THREAD

**Discussion by alexanderson1993** [Thorium-Sim/thorium-nova#161] [Feature Proposals]
There's a lot of stuff that the ship map enables, but we need a way to create those maps in the first place.

Ship maps are configured as part of a ship template, and consist of three parts.

### SVG layers

Each deck will first be defined with an SVG file that shows a graphical layout of that deck, with one SVG for each deck. Any images, lines, or diagrams can be used in these SVGs, but there will be guidelines for making sure the deck diagrams have a consistent appearance. SVGs are preferred since you can infinitely scale them without losing resolution, but it could be configured to work with transparent PNGs too.

As far as scale is concerned, we'll treat one SVG point as one meter. The expectation is that the deck SVGs will be able to layer on top of each other so they have everything lined up properly between decks.

In the future, there can be an in-app SVG editor, but for 1.0, we'll expect ship map SVGs to be created by hand or using an image editor like Adobe Illustrator.
So, decks will consist of:
- A name
- An order
- An SVG image

### Graph Nodes / Rooms

The purpose of the deck map is to allow entities to travel throughout the ship. Knowing where to move can't be done with an SVG alone. A graph of nodes is also required. Nodes represent places that these entities can travel, including junctions in hallways, turbo lifts for traveling between decks, and rooms.

Nodes are overlayed on top of the SVG layers and created using an in-app editor, which is described below.

Rooms are special nodes which have some extra features.

Rooms can be named, which lets the crew see what the room is for, have the room included in a list, and let the crew search for rooms by name.

Rooms can also be given icons, which are PNGs or SVGs. Rooms without icons will have a generic icon.

Rooms are assigned a radius. This represents the relative area that the room occupies on the ship map. When a crew member is inside a room, they'll be able to move about the room within that radius which makes it appear more realistic.

Finally, rooms have system assignments, which connects that room with a specific system. For example, cargo can be transferred between rooms. Rooms with the "probe launcher" assignment will allow probe casings in that room to be used for probe construction. Multiple systems can be assigned. Rooms without specific icons might be able to use an icon associated with the assigned system.

### Edges

Edges complete the graph by connecting nodes together. They can connect nodes on the same deck, or connect nodes between decks. Only one edge can connect two nodes, but nodes can have many edges. 

Edges can be given weights, which change the speed that entities traverse that edge by multiplying the entities speed by the weight. By default, the weight is 1. Higher weights make the entities move slower; lower weights move entities faster.

## Editing the Ship Map

Creating and editing the ship map is a simple process:

- First, create and name a deck and upload an image for it. The image is stored in the file system with the plugin, not in the .yml file for the ship template.
- Next, nodes are added. An editor palette allows configuring the node properties, and the node can be dragged around to change its position. 
- Finally edges are added between nodes. Clicking and dragging with the edge tool allows edges to easily be created. A "Add Edge" button allows edges to be created between decks without clicking and dragging. Edges can be selected to change the weight.

## What this enables:

- Cargo containers for transferring cargo
- Crew members, who operate as AI agents, with motivations and objectives
- Utility maps for power, water, heat, etc.
- A* path finding for any moving entities.

## What to comment on

If you see anything amiss about this, or have better ideas, let me know. Also, if you have ideas for things that can be enabled by these features, comment on those below. If there are no objections within a week or so, I'll start creating feature issues for the tasks needed to create this.

**Comment by Froginator45** [comment]
With the edges only being able to connect to 2 nodes, would this mean that you would need to connect every node to one another in a giant spiderweb for AI pathing? Or would the AI path via node chaining (without going into each node)? 

Also, would there be different edges for the different things (AI pathing, cargo, utility maps)? With a layer filter in the editor?

If I remember, originally I thought you were leaning towards the edges done via different line colors/weights in the SVG. I can see that being both a pain and useful. Is that idea being replaced with the in-app editor?

For the rooms, I can see the radius for random pathing being easy to program but is there any concern for oddly shaped rooms?

Something else to consider is a ship layout without corridors (i.e. rooms oriented around a "great hall").

**Reply by alexanderson1993** [reply]
Edges only connecting to two nodes just means you can't have duplicate edges between two nodes, and all edges are bi-directional. I don't see any reason to have directed edges. The AI pathfinding would find the shortest path to get where it wants to go, though I haven't decided how weights play into that.

Different edges for different things makes sense, perhaps, for crew vs cargo, however utility maps will be another layer separate from the nodes/edges discussed here. For the initial implementation, we'll have a unified graph for crew and cargo. Separating it into separate graphs later won't be terribly hard to refactor.

> different line colors/weights in the SVG

I vaguely remember this suggestion, but it has a lot of problems, specifically with using really strange conventions that have to be carefully documented and implemented. Drawing the nodes and edges in-app (with the implication that they can also be manually written in the YAML file) seems more straightforward.

>  is there any concern for oddly shaped rooms?

Yeah, this is a weird one that I think should be addressed _some time_. `radius` is good enough for the first pass, and we can look into giving nodes shapes in a future iteration. 

>  a ship layout without corridors

I imagine this would still work fine with this model. Just have several interconnected nodes within the great hall in a multi-layer hub and spoke model. Not exactly ideal, but it would work good enough.

Great thoughts and feedback! I just want to emphasize that for many of these I'm saying "not yet, but yes, eventually". Push back if I've dismissed any of them too quickly.

**Reply by Froginator45** [reply]
The only thing I can think of using directed edges would be utilities. But they could be simplified to not need direction. Direction would only be nice if you were trying to make a more realistic setup for more advanced players. Maybe this could be a feature?

**Reply by alexanderson1993** [reply]
I can see that, but utilities are outside the scope of this proposal, so we'll cover that when we start working on Utilities.

**Comment by astrolamb-gaming** [comment]
So would a single node represent a single room/hallway section? And the SVG is really just a visual representation of the rooms, etc?

**Reply by alexanderson1993** [reply]
> So would a single node represent a single room/hallway section?

A hallway would be comprised of several nodes, one at every intersection or doorway. I imagine we could achieve abnormal room layouts by placing  two identical room nodes right next to each other and programming the crew to move between them if they have a single edge between them, to simulate wandering around a room. 

Here's a diagram I just drew that hopefully demonstrates this better. Red is nodes, blue is edges. Notice how all of the edges go around and towards the circular turbo lift, so crew can enter from any direction.

![Image](https://user-images.githubusercontent.com/6558157/148291250-49cdc268-15c2-4717-a76b-bcb093150f45.jpeg)

> the SVG is really just a visual representation of the rooms, etc?

Yep, I think the SVG/Image will just be for looks. I thought about having more complicated behavior by assigning IDs to the SVG elements, but that very quickly makes it difficult to create and edit.
